### PR TITLE
nixos/switch-to-configuration: Add a massive test and do a slight refactor

### PR DESCRIPTION
**Pull request by dasJ**
###### Motivation for this change

To make my next change easier I refactored the handling of modified units into a function and also added a lot of test cases to the switch test.
I also changed non-service units to be restarted instead of being started and stopped because it makes no sense to stop and start units that cannot have `ExecStop=`

###### Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests) → switchTest
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: nixos/modules/system/activation/switch-to-configuration.pl
Comment (lovesegfault) [inline nixos/modules/system/activation/switch-to-configuration.pl:163]: Maybe add a comment here on why we're not doing anything currently

**Review by lovesegfault — APPROVED** [review]
Looks good!

**Comment by primeos** [comment]
@dasJ thanks for splitting these changes up into smaller PRs! :) I just want to drop by a brief comment that I'm not ignoring them and intend to review them during the weekend (unfortunately I don't have enough time for a proper review during the week - if it's fine with you I'd like to delay merging until then but I can also retroactively review it like I'll already have to do with #148751).

**Comment by dasJ** [comment]
@primeos I didn't wait for your review because I was sprinting the changesets and I plan on continuing that. After this PR there is another in the pipeline that will touch up on the "restart by activation script" logic (and maybe yet another small one?) and then I will get to the final PR that restarts sockets. This will be the one I won't push people about merging and instead give you (and others) some time to test and review it. I am a lot more comfortable now because I was able to find the mentioned issues in the code and I talked my new approach through with some committers. The next revision of this functionality will have less edge cases and have a test suite that is able to catch the behaviour that was previously broken.

**Review by primeos — APPROVED** [review]
Huge thanks for the extended VM test and adding it to `systemd.passthru.tests`! <3

I only briefly reviewed the `handleModifiedUnit` extraction into a subroutine via `git show --color-moved --color-moved-ws=allow-indentation-change 393c72184986e66f2e72cf0d7e6c0476447c10b6` but that looks functionally identical to me (as expected/intended) so it should be fine.

I didn't properly review the extended Python test script yet (I'll take a look later out of interest) but that shouldn't really be able to make anything worse anyway :)

So +1 and thanks from me.
CI successfully ran the switchTest: /nix/store/bkbgsj2lv7y5bawyccswjs0r404cz3nq-vm-test-run-switch-test

File: nixos/modules/system/activation/switch-to-configuration.pl
Comment (dasJ) [reply nixos/modules/system/activation/switch-to-configuration.pl:178]: Yeah I refactored the logic in the old PR and I plan on doing the same this time again but I wanted the code to be as unmodified as possible for this PR

**Pull request merged**
